rfcs/20251203-grouped-gemm-support/README.md

```diff
@@ -0,0 +1,604 @@
+
+(**) Regrouping involves gathering all tokens assigned to each expert
+and creating a contiguous block of memory for each expert's input.
+
+Grouped GEMM processes variable-sized batches (subsets of tokens) per expert,
+which is a key difference from batched GEMM operations that assume uniform batch sizes across groups.
+
+For each expert `i` (dimensions are: `src[Mi x K]`, `weights[K x N]`, `dst[Mi x N]`) we need to compute:
+```
+output[i] = (src[i] * weights[i]) + bias[i]
+```
+Plus scaling factors, zero-point adjustments, etc. (as in regular matrix multiplication).
+
+To reiterate **terms used/considerations of MoE:**
+- **Number of experts**: Total number of experts in the MoE layer.
+- **TOP-K routing**: A routing strategy where each token is assigned to the top K experts according to some scoring mechanism.
+- **Number of active experts**: Number of experts that received tokens after routing.
+- **M dimension is Dynamic dimension**: M dimension (i.e., token count per expert) varies dynamically based on routing decisions.
+And depending on the framework design, this information may only be available on the device side after routing.
+- **Uniform K and N**: All experts share the same K and N dimensions, so weight shapes are the same.
+- **Uniform scaling configurations**: All experts use the same scaling pattern (e.g., per-tensor, per-row), but different values.
+
+See also:
+- [Useful resource to learn about MoE in PyTorch with examples of MoE workflow and challenges](https://pytorchconference.sched.com/event/27QE0/pytorch-apis-for-high-performance-moe-training-and-inference-daniel-vega-myhre-ke-wen-natalia-gimelshein-meta)
+
+## Frameworks Implementations
+
+### PyTorch Grouped Scaled MM
+
+PyTorch provides [`torch/nn/functional.py`](https://github.com/pytorch/pytorch/blob/main/torch/nn/functional.py) API for grouped scaled matrix multiplication.
+The `torch._scaled_grouped_mm` API is used internally:
+
+```python
+torch._scaled_mm_grouped(
+    mat_a,                           # Input tensor A (concatenated groups)
+    mat_b,                           # Input tensor B (concatenated groups)
+    scale_a,                         # Scaling factors for mat_a
+    scale_recipe_a,                  # Scaling type for mat_a, choises are: PER_TENSOR, PER_ROW, PER_COLUMN, AXISWISE
+    scale_b,                         # Scaling factors for mat_b
+    scale_recipe_b,                  # Scaling type for mat_b
+    swizzle_a=None,                  # Swizzle pattern for scale_a
+    swizzle_b=None,                  # Swizzle pattern for scale_b
+    bias=None,                       # Optional bias tensor
+    offs=None,                       # Device-side offsets for source tensor
+    output_dtype=torch.bfloat16,     # Output data type
+    contraction_dim=None,            # Which dims are K in matmul
+    use_fast_accum=False             # Nvidia GPUs specifics
+)
```

Comment:
This API is internal right? Do you know how does it plug to PyTorch external API?
If this gets called only from Inductor pattern matcher, do we have to align to this internal API or is it ok to deviate?